`PolyMatrix` class for a matrix of polynomials

gxyd opened the pull request:
Add doesn't return Poly1 for Add(Poly1, Poly(0, x))

In master
```
>>> import sympy
>>> sympy.__version__
'1.0.1.dev'
>>> from sympy import *
>>> x = Symbol('x')
>>> Mq = Matrix([[Poly(x**2, x, domain='QQ'), Poly(0, x, domain='QQ'), Poly(-x**2, x, domain='QQ'), Poly(x**3, x, domain='QQ'), Poly(0, x, domain='QQ'), Poly(-x**3, x, domain='QQ')]])
>>> vj = Matrix([1, 0, 0, 0, 0, 0])
>>> Mq*vj
Matrix([[Poly(x**2, x, domain='QQ') + 5*Poly(0, x, domain='QQ')]])
```

<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->

asmeurer commented:
@siefkenj

asmeurer commented:
`sum` is theoretically slower than Add, because it would call flatten for each sub-sum (quadratic time vs. linear time).

asmeurer commented:
From discussion on gitter, I would rather see a Matrix subclass that directly works with Poly elements (or polys domain elements). How hard would that be to do, @siefkenj @jksuom?

siefkenj commented:
`Add` is faster than `sum`, which is why it's used there.  It should be fairly easy to subclass `DenseMatrix` to make a `PolyDenseMatrix`.  Just override the appropriate `_eval_*` functions to make them poly-specific.

jksuom commented:
> subclass DenseMatrix to make a PolyDenseMatrix

That is my suggestion as well, The abbreviated name could be `PolyMatrix`. There remains to decide where it should be put. Perhaps `polyutils`.

gxyd committed:
Add PolyMatrix class

gxyd commented:
I am guessing there would be a need to make some changes in the `classof` function as well?

jksuom commented:
`classof` checks the class priority. That would probably suffice.

gxyd commented on sympy/polys/polymatrix.py:
```diff
@@ -0,0 +1,69 @@
+    Matrix([[Poly(x**2, x, domain='QQ')]])
+
+    """
+    _class_priority = 5
```
Also do you think that this value would be good? `_class_priority` for `ImmutableMatrix` is 8.

siefkenj commented on sympy/polys/polymatrix.py:
```diff
@@ -0,0 +1,69 @@
+                vec = (mat[a]*other_mat[b] for a,b in zip(row_indices, col_indices))
+                new_mat[i] = sum(vec)
+
+        return classof(self, other)._new(new_mat_rows, new_mat_cols, new_mat, copy=False)
```
Unless you're planning on having a `SparsePolyMatrix` you can just do `self._new`

siefkenj commented on sympy/polys/polymatrix.py:
```diff
@@ -0,0 +1,69 @@
+        return cls._new(*args, **kwargs)
+
+    @classmethod
+    def _new(cls, *args, **kwargs):
```
I don't think you need to redefine this method, unless you were going to store special things, like the poly domain with the object.

siefkenj commented on sympy/polys/polymatrix.py:
```diff
@@ -0,0 +1,69 @@
+    """
+    _class_priority = 5
+
+    def __new__(cls, *args, **kwargs):
```
I don't think you need to redefine this method

gxyd replied:
Docstring of `DenseMatrix` says:
```
A class to house deprecated matrix methods.
```
Also if I remove the `__new__` method from this (as you suggest), I get an error with the creation of object.
```
>>> Mq = PolyMatrix([[Poly(x**2, x, domain='QQ')]])
TypeError                                 Traceback (most recent call last)
<ipython-input-16-94e2171a82db> in <module>()
----> 1 Mq = PolyMatrix([[Poly(x**2, x, domain='QQ')]])

TypeError: object() takes no parameters
```
Then shouldn't this have `__new__` method just like `MutableDenseMatrix` or `ImmutableDenseMatrix` have theirs?

siefkenj replied:
Interesting note on the docstring.  That should be overridden.  I see your point now.  Probably, there should be a `PolyMatrix` which gets subclassed as a `ImmutablePolyMatrix` and `MutablePolyMatrix` and the `Mutable`/`Immutable` versions have the `__new__` method.

jksuom commented:
For now, it should be enough to redefine `_eval_matrix_mul`. To make this usable more generally for `polys` objects, also `_sympify` should be redefined (to the trivial `staticmethod(lambda x: x)`).

asmeurer commented:
Is it much more work to make this operate on the Poly domain directly instead of on the Poly (so that the entire matrix has a unified domain)? If so, we should put it off, but I think that is what we eventually want to have.

jksuom commented:
It is possible that redefining `_sympify` will suffice. That is what is done in `polys.solvers`.

asmeurer commented:
> It is possible that redefining _sympify will suffice. That is what is done in polys.solvers.

So maybe it is easy to make it so that you call `PolyMatrix(..., domain=QQ)` and the `_sympify` method basically becomes `domain.unify`. Getting the domain automatically is a bit more work but probably not that hard (just need to use whatever function Poly uses). 

Note that a very key advantage of making the entries be from the domain instead of Poly is that the matrix can actually contain elements from a field, without having to "fake" it with a polynomial in an unrelated variable, since don't have [Frac](https://github.com/sympy/sympy/issues/5141) objects.

asmeurer commented:
> the `_sympify` method basically becomes `domain.unify`

Or actually I guess just `domain`.

gxyd committed:
Fix docstring and change class priority of _class_priority

The priority should probably be higher than that of any standard matrix.
Some version of PolyMatrix should be returned.

gxyd committed:
classof -> self and add a doctest

gxyd commented on sympy/polys/polymatrix.py:
```diff
@@ -0,0 +1,73 @@
+    >>> v1*pm1
+    Matrix([
+    [ Poly(x**2, x, domain='ZZ'),  Poly(-x, x, domain='ZZ')],
+    [-Poly(x**2, x, domain='ZZ'), -Poly(-x, x, domain='ZZ')]])
```
This is not correct, since `S(-1)*Poly(-x, x)` returns `-Poly(-x, x)` while `Poly(-x, x)*S(-1)` returns `Poly(x, x)`.

gxyd replied:
This is the main trouble, as `-Poly(-x, x)` is of type `Mul` and `.degree` method will raise an `AttributeError`.

gxyd committed:
add _sympify routine to PolyMatrix

jksuom commented on sympy/polys/polymatrix.py:
```diff
@@ -0,0 +1,74 @@
+
+class PolyMatrix(DenseMatrix):
+    """
+    Matrix of polynomials.
```
There could also be other polys' objects, e.g., fractions. I'd think of 'Poly` as referring to everything in `polys`.

jksuom commented on sympy/polys/polymatrix.py:
```diff
@@ -0,0 +1,74 @@
+    _class_priority = 10
+    _sympify = staticmethod(lambda x: x)
+
+    def __new__(cls, *args, **kwargs):
```
I don't that this should be redefined. Trivial `_sympify` will make it possible to have non-sympifiable object in these matrices even with the standard `__new__`. If something else need be added, that could probably be done in a custom `__init__`.

gxyd replied:
@siefkenj suggested the same thing, for which I commented https://github.com/sympy/sympy/pull/12694#discussion_r119388271. Should we subclass `MutableDenseMatrix` not `DenseMatrix`?

jksuom replied:
I think it should be the same as `Matrix`, which is an abbreviation of `MutableDenseMatrix`:
```
In [1]: Matrix?
Type:       type
Base Class: <type 'type'>
String Form:<class 'sympy.matrices.dense.MutableDenseMatrix'>
Namespace:  Interactive
File:       /home/jks/sympy/sympy/matrices/dense.py
Docstring:  <no docstring>
```

gxyd replied:
Yes, which it is right now. I verbatim copied the code from `MutableDenseMatrix`.

gxyd committed:
Remove unused import

gxyd committed:
Add tests file for polymatrix.py

gxyd commented:
I have added the test file, it seems like adding `PolyMatrix` to the `polyutils.py` (as you had suggested) instead of making a new file, would require making a lot of inner (inside function) imports because of circular import. But I will still try to see if it is possible.

gxyd committed:
PolyMatrix subclassed as MutablePolyMatrix, ImmutablePolyMatrix

gxyd commented:
I have sub-classed it further as Mutable/Immutable `PolyMatrix` counterparts of `DenseMatrix`.

Also seems like the tests passed in the last commit (someone has to restart them though it seems).

jksuom commented:
I'm not sure that an immutable PolyMatrix would be needed. If we want to transform a PolyMatrix to a MatrixExpr with Basic entries, I think we could use the standard ImmutableMatrix.

gxyd committed:
Add row_op and col_op methods and remove ImmutablePolyDenseMatrix

If we want to transform a PolyMatrix to a MatrixExpr with Basic entries,
I think we could use the standard ImmutableMatrix.

gxyd commented:
I have added `row_op` and `col_op` methods, as they are going to be needed in any case as well.
For the removal of `ImmutablePolyDenseMatrix`, I have removed it, your argument sounds alright.

jksuom commented:
> I have added row_op and col_op methods,

We could derive PolyMatrix from MutableDenseMatrix. Then these methods would be inherited.

gxyd committed:
Subclass PolyMatrix from MutableDenseMatrix instead of DenseMatrix

gxyd commented:
I have removed the `[WIP]` from the title.

jksuom commented:
This looks good. It can be expanded later if that should prove desirable.
There could also be a test for the product of a PolyMatrix with an ordinary matrix.

gxyd committed:
Add _eval_scalar_mul and _eval_scalar_rmul methods

Also add tests

gxyd commented on sympy/polys/tests/test_polymatrix.py:
```diff
@@ -0,0 +1,39 @@
+
+    pm2 = PolyMatrix([[Poly(x**2, x, domain='ZZ'), Poly(-x**2, x, domain='ZZ')]])
+    v2 = Matrix([1, -1])
+    assert pm2*v2 == PolyMatrix([[Poly(2*x**2, x, domain='ZZ')]])
```
I don't think we make `v2*pm2` to evaluate such that it returns containing `Poly(x**2, x)` instead of `-Poly(-x**2, x)`.

siefkenj commented:
I disagree with only having `PolyDenseMatrix`. The eventual plan is to have the default matrix type be `ImmutableDenseMatrix`.  Since most sympy objects are immutable, this will eliminate a lot of confusion.  Further, I don't think you should implement `row_op` and `col_op` naively.  Since `PolyDenseMatrix` should be over the ring of polynomials, we shouldn't allow operations that could end up dividing by a polynomial without changing the matrix type.  Specifically, we cannot assume that we may "divide" by a non-zero entry of a `PolyDenseMatrix`.

Here are some cases we need to decide:

    m = PolyMatrix([[Poly(2*x**2, x, domain='ZZ')]])

    .5*m
    m / Poly(2*x, x, domain='ZZ')
    m / Poly(x**3, x, domain='ZZ')

In both cases the result will no longer have domain `'ZZ'`.  In the last case, the result cannot be a `PolyMatrix` at all.  Should some of these cases throw errors?  Should these be up-cast to a domain where they make sense?  I don't work in non-field rings that often, so I don't know what would be most useful, but I would think that if someone asked for an object over a particular ring, they might like it to warn them if they left the ring.

If the answer is they should be up-cast, the detection can be done in `PolyMatrix._new`.  Since this is called for all non-mutating operations, it would be an easy fix.  However, if you allow a matrix to be mutated, you have to throw an error if the user mutates an entry outside the desired ring!

jksuom commented:
> we shouldn't allow operations that could end up dividing by a polynomial without changing the matrix type

A lot of type changes occur silently in polys. But those should probably be avoided in this case. So some checks could be included when the ring attribute has been added to PolyMatrix.

> The eventual plan is to have the default matrix type be ImmutableDenseMatrix.

Immutable matrices do currently expect that their entries are instances of Basic. That will not be possible for polys objects. If that restriction can lifted, then PolyMatrix might be derived from ImmutableDenseMatrix. I don't think that mutability itself will be necessary.

gxyd committed:
remove failing test and merge test functions

jksuom commented:
I think this is usable already and could be merged even without those checks mentioned above. (This is experimental and the caller would be responsible for proper use.) One thing could be added though, the object attribute `ring`. That could be initialized by an `__init__` method that would determine the ring by scanning `self._ mat`.

gxyd committed:
Add __init__ magic method and add .ring method to PolyMatrix

For a PolyMatrix having any non-Poly element, then the ring **kwarg
argument must be provided to it

gxyd commented on sympy/polys/polymatrix.py:
```diff
@@ -0,0 +1,77 @@
+
+from sympy.matrices.dense import MutableDenseMatrix
+from sympy.polys.polytools import Poly
+from sympy.core.compatibility import reduce
```
I think it wouldn't be a problem to use the `reduce` from SymPy instead of python's `reduce`.

jksuom replied:
SymPy's `reduce` is, in fact, Python's reduce. Only it has been moved from builtins to functools in Python 3. It seems that its use is discouraged in favour of an explicit loop construction.

jksuom commented:
This looks good. Luckily matrices accept all keyword arguments.

gxyd committed:
use 'for' loop instead of 'reduce' in PolyMatrix

gxyd committed:
remove unused import 'reduce'

gxyd commented on sympy/polys/tests/test_polymatrix.py:
```diff
@@ -0,0 +1,40 @@
+    v4 = Matrix([1, -1], ring='ZZ[x]')
+    assert pm4*v4 == PolyMatrix([[Poly(2*x**2, x, domain='ZZ')]])
+
+    assert len(PolyMatrix()) == 0
```
This used to raise an error with `reduce` (for obvious reasons), have now use `for` loop instead.

jksuom commented:
Ok, I think this is good for its intended purpose now.
